cache/sampleStoreTimeout.js

```diff
@@ -0,0 +1,94 @@
+/**
+ * Copyright (c) 2017, salesforce.com, inc.
+ * All rights reserved.
+ * Licensed under the BSD 3-Clause license.
+ * For full license text, see LICENSE.txt file in the repo root or
+ * https://opensource.org/licenses/BSD-3-Clause
+ */
+
+/**
+ * ./cache/sampleStoreTimeout.js
+ *
+ * Timeout samples
+ */
+'use strict'; // eslint-disable-line strict
+const sampleStore = require('./sampleStore');
+const redisClient = require('./redisCache').client.sampleStore;
+const isTimedOut = require('../db/helpers/sampleUtils').isTimedOut;
+const constants = require('../api/v1/constants');
+const ONE = 1;
+const TWO = 2;
+
+module.exports = {
+
+  /**
+   * Invalidates samples which were last updated before the "timeout" specified
+   * by the aspect. Get all samples and corresponding aspects. If the sample
+   * should be timed out, set sample value, status, previous status, status
+   * changed at and updated at fields.
+   * @param  {Date} now - Date object
+   * @returns {Promise} - Resolves to the number of evaluated and timed out
+   * samples
+   */
+  doTimeout(now) {
+    const curr = now || new Date();
+    let numberTimedOut = 0;
+    let numberEvaluated = 0;
+
+    return new Promise((resolve, reject) => {
+      redisClient.smembersAsync(sampleStore.constants.indexKey.sample)
+      .then((allSamples) => {
+        const commands = [];
+        const aspectType = sampleStore.constants.objectType.aspect;
+
+        allSamples.forEach((sampKey) => {
+          const aspectName = sampKey.split('|')[ONE];
+          commands.push(['hgetall', sampKey]); // get sample
+          commands.push(
+            ['hgetall', sampleStore.toKey(aspectType, aspectName)]
```

Comment:
I think this is going to be sending tons of duplicate requests for the same aspect. How about adding each sample's aspect to a Set first, to consolidate it down to a list of unique aspects?